src/pyodide.js

```diff
@@ -171,102 +161,127 @@ var languagePluginLoader = new Promise((resolve, reject) => {
+      if (n === 0) {
+        if (runDepResolve !== undefined) {
+          runDepResolve();
+          runDepResolve = undefined;
+        }
+      } else {
+        if (runDepResolve === undefined) {
+          runDepPromise = new Promise(r => runDepResolve = r);
+        }
       }
+    };
 
-      let packageList = Array.from(Object.keys(toLoad));
-      _messageCallback(`Loading ${packageList.join(', ')}`)
+    // Add a promise that rejects when any exceptions is thrown in the process
+    // of loading a script. The promise never resolves, and we combine it
+    // with other promises via Promise.race.
+    let windowErrorHandler;
+    let windowErrorPromise = new Promise((_res, rej) => {
+      windowErrorHandler = e => rej(e.message);
+      self.addEventListener('error', windowErrorHandler);
+    });
 
-      // monitorRunDependencies is called at the beginning and the end of each
-      // package being loaded. We know we are done when it has been called
-      // exactly "toLoad * 2" times.
-      var packageCounter = Object.keys(toLoad).length * 2;
+    // Promises for each script load. The promises already handle error and
+    // never fail.
+    let scriptPromises = [];
 
-      self.pyodide._module.monitorRunDependencies = () => {
-        packageCounter--;
-        if (packageCounter === 0) {
-          for (let pkg in toLoad) {
-            self.pyodide.loadedPackages[pkg] = toLoad[pkg];
-          }
-          delete self.pyodide._module.monitorRunDependencies;
-          self.removeEventListener('error', windowErrorHandler);
-
-          let resolveMsg = `Loaded `;
-          if (packageList.length > 0) {
-            resolveMsg += packageList.join(', ');
-          } else {
-            resolveMsg += 'no packages'
-          }
+    for (let pkg in toLoad) {
+      let scriptSrc;
+      let package_uri = toLoad[pkg];
+      if (package_uri == 'default channel') {
+        scriptSrc = `${baseURL}${pkg}.js`;
+      } else {
+        scriptSrc = `${package_uri}`;
+      }
+      console.log(`Loading ${pkg} from ${scriptSrc}`)
+      scriptPromises.push(loadScript(scriptSrc).catch(() => {
+        console.error(`Couldn't load package from URL ${scriptSrc}`);
+        delete toLoad[pkg];
+      }));
+    }
 
-          if (!isFirefox) {
-            preloadWasm().then(() => {
-              console.log(resolveMsg);
-              resolve(resolveMsg);
-            });
-          } else {
-            console.log(resolveMsg);
-            resolve(resolveMsg);
-          }
-        }
-      };
+    // Wait for all scripts to be loaded first. Then all the run dependencies
+    // have been added. Then wait for the dependencies to be removed.
+    let successPromise = Promise.all(scriptPromises).then(() => runDepPromise);
+    try {
+      await Promise.race([ successPromise, windowErrorPromise ]);
+    } finally {
+      delete self.pyodide._module.monitorRunDependencies;
+      self.removeEventListener('error', windowErrorHandler);
```

Comment:
Both of these lines make me really unhappy (I know this is just stuff you moved around).

Reply:
I'm open to suggestions, but otherwise I'm satisfied that I have left the code in a slightly nicer state.